src/mlpack/methods/ann/layer/bilinear_interpolation_impl.hpp

```diff
@@ -22,168 +23,198 @@ namespace ann /** Artificial Neural Network. */ {
+        coeffs_pre.slice(current_idx)(0, 0) = (1 - deltaC) * (1 - deltaR);
+        coeffs_pre.slice(current_idx)(0, 1) = (1 - deltaR) * deltaC;
+        coeffs_pre.slice(current_idx)(1, 0) = deltaR * (1 - deltaC);
+        coeffs_pre.slice(current_idx)(1, 1) = deltaR * deltaC;
+        // Bottom Right corner of the Pi's for each point in Output matix
+        index_pre.slice(current_idx)(0, 0) = c_end + 1;
+        index_pre.slice(current_idx)(0, 1) = r_end + 1;
+      }
+    }
+  }
 
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BilinearInterpolation<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
-{
-  batchSize = input.n_cols;
-  if (output.is_empty())
-    output.set_size(outRowSize * outColSize * depth, batchSize);
-  else
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
   {
-    assert(output.n_rows == outRowSize * outColSize * depth);
-    assert(output.n_cols == batchSize);
-  }
+    batchSize = input.n_cols;
+    if (output.is_empty())
+      output.set_size(outRowSize * outColSize * depth, batchSize);
+    else
+    {
+      assert(output.n_rows == outRowSize * outColSize * depth);
+      assert(output.n_cols == batchSize);
+    }
 
-  assert(inRowSize >= 2);
-  assert(inColSize >= 2);
+    assert(inRowSize >= 2);
+    assert(inColSize >= 2);
 
-  arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+    arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
       inRowSize, inColSize, depth * batchSize, false, false);
-  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
-                          depth * batchSize, false, true);
-
-  double scaleRow = (double) inRowSize / (double) outRowSize;
-  double scaleCol = (double) inColSize / (double) outColSize;
+    arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
+      depth * batchSize, false, true);
 
-  arma::mat22 coeffs;
-  for (size_t i = 0; i < outRowSize; ++i)
-  {
-    size_t rOrigin = (size_t) std::floor(i * scaleRow);
-    if (rOrigin > inRowSize - 2)
-      rOrigin = inRowSize - 2;
-
-    // Scaled distance of the interpolated point from the topmost row.
-    double deltaR = i * scaleRow - rOrigin;
-    if (deltaR > 1)
-      deltaR = 1.0;
-    for (size_t j = 0; j < outColSize; ++j)
+    for (size_t k = 0; k < depth * batchSize; ++k)
     {
-      // Scaled distance of the interpolated point from the leftmost column.
-      size_t cOrigin = (size_t) std::floor(j * scaleCol);
-      if (cOrigin > inColSize - 2)
-        cOrigin = inColSize - 2;
-
-      double deltaC = j * scaleCol - cOrigin;
-      if (deltaC > 1)
-        deltaC = 1.0;
-      coeffs[0] = (1 - deltaR) * (1 - deltaC);
-      coeffs[1] = deltaR * (1 - deltaC);
-      coeffs[2] = (1 - deltaR) * deltaC;
-      coeffs[3] = deltaR * deltaC;
-
-      for (size_t k = 0; k < depth * batchSize; ++k)
+      // Padded Input Matrix
```

Comment:
Is this some special padding here, or just top, bottom, left, right? Can you add some comments here to make this more clear?